stubs/torch/nn/__init__.pyi

```diff
@@ -0,0 +1,5 @@
+from .modules import *
```

Comment:
I am curious to know - Do we simply copy-paste these interface `.pyi` files from typeshed repository?

Reply:
Basically, yes. The official stubs for `nn.Module`, `nn.ModuleList`, `nn.ParameterList`, and `nn.Parameter` are simply incorrect, so I had to provide my own.

The reason why I also had to copy a bunch of files is that mypy uses different import mechanisms. I'm not entirely sure what happens behind the hood, but from my experiments I guess what happens is:

- mypy scans for stub files under `site-packages` and `mypy_path` (defined in `mypy.ini`)
  - Stubs in `site-packages` are put beside their code files (i.e., file with the same name under the same path, except the extension's changed to `.pyi`).
  - Stubs in `mypy_path` should follow the same folder structure as those in `site-packages` (in our case, e.g., `stubs/torch/nn/parameter.pyi` corresponds to the `torch.nn.parameter` package).
- Stubs in `mypy_path` are scanned first. `site-packages` ones are used if no corresponding stubs exist in `mypy_path`.
- If an import in `mypy_path` stub does not exist (e.g., `.parallel` and `init` here), stubs in `mypy_path` are imported.
  - However, this does not work the other way around. `site-packages` stubs always import from `site-packages`, even if corresponding stubs exist in `mypy_path`.
- If a named is previously defined, it will not be overwritten by later definitions.

These observations are exactly why we have to copy-paste these `.pyi` files:

- We cannot only add stubs for `nn.Module` and classes, and import them afterwards to overwrite the official stubs.
- We have to put these stubs under the same structure, e.g. `container.pyi` under `stubs/torch/nn/modules`.
- We have to write our own `__init__.pyi` for each layer of package, although the contents are exactly the same as the official stubs.

Reply:
Great explanation, thanks.